### PR TITLE
feat: add enabled/blacklist fields to tokens

### PR DESCRIPTION
**Pull request by toteki**
## Description

Adds various options for disabling borrow, lend, and oracle price calculations of asset types.

Prevents token registry update from EVER removing base denoms from token registry. Therefore:
- Gov updates only need to include the tokens they're actually adding/changing.
- Removing tokens intentionally can be done by disabling borrowing/lending, or setting the `Blacklist` field depending on the situation. Their metadata and params remain in the token registry.

Also adds validation to `SetRegisteredToken`

closes: #869
closes: #724 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/gov_handler.go [hunk 1]
```diff
@@ -22,12 +22,10 @@ func NewUpdateRegistryProposalHandler(k keeper.Keeper) govtypes.Handler {
...
-	if err := k.DeleteRegisteredTokens(ctx); err != nil {
-		return err
-	}
-
 	for _, token := range p.Registry {
-		k.SetRegisteredToken(ctx, token)
+		if err := k.SetRegisteredToken(ctx, token); err != nil {
+			return err
+		}
 	}
```
Comment (toteki) [inline x/leverage/gov_handler.go:29]: Note the functional difference in update token registry - now any existing tokens will remain or be updated, instead of being deleted. It is no longer possible to delete registered tokens entirely.

File: x/leverage/keeper/oracle.go [hunk 2]
```diff
@@ -14,11 +14,16 @@ import (
 // the base and display/symbol denominations for each exchange pair. E.g. it must
 // know about the UMEE/USD exchange rate along with the uumee base denomination
 // and the exponent.
-// This function will not return non-positive exchange rates, preferring
-// to error instead.
+// This function will not return non-positive exchange rates unless a token is
+// blacklisted, in which case it will return zero.
```
Comment (toteki) [inline x/leverage/keeper/oracle.go:18]:
```suggestion
// This function will only return positive exchange rates or errors, unless a
// token is blacklisted, in which case it will return zero.
```

File: x/oracle/keeper/hooks_test.go [hunk 3]
```diff
@@ -9,6 +9,12 @@ func (s *IntegrationTestSuite) TestHooks_AfterTokenRegistered() {
...
 	s.Require().Len(s.app.OracleKeeper.AcceptList(s.ctx), 1)
 
+	// TODO: These hooks need to respond to the Token.Blacklist field
+	// (as well as update symbol denom and exponent if they ever change).
+	// Blacklisting in particular should actually eliminate the base
+	// denom from the oracle (and un-blacklisting, for whatever reason,
+	// should do the opposite.)
+
 	// require that an existing token does not change the accept list
 	h.AfterTokenRegistered(s.ctx, leveragetypes.Token{
```
Comment (toteki) [inline x/oracle/keeper/hooks_test.go:19]: @adamewozniak Could you add the necessary oracle changes to this PR after reviewing?

File: x/leverage/gov_handler_test.go [hunk 4]
```diff
@@ -30,30 +30,93 @@ func TestUpdateRegistryProposalHandler(t *testing.T) {
...
 
+		// ensure that unaffected token was not deleted by the gov proposal
 		token, err := k.GetRegisteredToken(ctx, "uatom")
 		require.NoError(t, err)
+
+		// ensure that the existing token was updated by the gov proposal
+		token, err = k.GetRegisteredToken(ctx, "uosmo")
+		require.NoError(t, err)
 		require.Equal(t, "0.020000000000000000", token.BaseBorrowRate.String())
 	})
```
Comment (robert-zaremba) [inline x/leverage/gov_handler_test.go:121]: let's add a test checking that a token with `enableLend=true, blacklisted=true` will fail to process - according to proto, when blacklisted=true then enable* must be false.

File: x/oracle/keeper/hooks_test.go (hunk 3 above)
Comment (adamewozniak) [reply x/oracle/keeper/hooks_test.go:19]: @toteki I don't believe this update is necessary; the oracle module should be relatively agnostic of the x/leverage module's functionality. 

For example, if we used the oracle as a service for other blockchains over IBC, we would not want a leverage "blacklist" to remove those from the oracle's params.

In the case where we **definitely do** want to remove a denom from the oracle, a parameter governance proposal is still possible.

Comment (adamewozniak) [reply x/oracle/keeper/hooks_test.go:19]: @toteki I don't believe this update is necessary; the oracle module should be relatively agnostic of the x/leverage module's functionality. 

For example, if we used the oracle as a service for other blockchains over IBC, we would not want a leverage "blacklist" to remove those from the oracle's params.

In the case where we **definitely do** want to remove a denom from the oracle, a parameter governance proposal is still possible.

File: x/leverage/keeper/keeper_test.go [hunk 5]
```diff
@@ -618,8 +643,13 @@ func (s *IntegrationTestSuite) TestLiqudateBorrow_Valid() {
 		MaxBorrowRate:        sdk.MustNewDecFromStr("1.0"),
 		KinkUtilizationRate:  sdk.MustNewDecFromStr("0.8"),
 		LiquidationIncentive: sdk.MustNewDecFromStr("0.1"),
+		SymbolDenom:          "UMEE",
+		Exponent:             6,
+		EnableLend:           true,
+		EnableBorrow:         true,
+		Blacklist:            false,
 	}
```
Comment (robert-zaremba) [inline x/leverage/keeper/keeper_test.go:651]: let's try to reduce copy pasted definitions and define them once in a suite.

File: x/leverage/gov_handler_test.go [hunk 6]
```diff
@@ -30,30 +30,93 @@ func TestUpdateRegistryProposalHandler(t *testing.T) {
...
 
 		tokens := k.GetAllRegisteredTokens(ctx)
-		require.Len(t, tokens, 2)
+		require.Len(t, tokens, 3)
 
-		_, err := k.GetRegisteredToken(ctx, "uosmo")
-		require.Error(t, err)
-
-		_, err = k.GetRegisteredToken(ctx, "uumee")
+		// ensure that new tokens was added
```
Comment (adamewozniak) [inline x/leverage/gov_handler_test.go:109]:
```suggestion
```

File: x/leverage/gov_handler_test.go [hunk 7]
```diff
@@ -30,30 +30,93 @@ func TestUpdateRegistryProposalHandler(t *testing.T) {
...
+		_, err := k.GetRegisteredToken(ctx, "uumee")
 		require.NoError(t, err)
 
+		// ensure that unaffected token was not deleted by the gov proposal
```
Comment (adamewozniak) [inline x/leverage/gov_handler_test.go:113]:
```suggestion
		// ensure that only the specified token was deleted
```

File: x/leverage/gov_handler_test.go [hunk 8]
```diff
@@ -30,30 +30,93 @@ func TestUpdateRegistryProposalHandler(t *testing.T) {
...
 		token, err := k.GetRegisteredToken(ctx, "uatom")
 		require.NoError(t, err)
+
+		// ensure that the existing token was updated by the gov proposal
```
Comment (adamewozniak) [inline x/leverage/gov_handler_test.go:117]:
```suggestion
```

File: x/leverage/keeper/keeper.go [hunk 9]
```diff
@@ -67,8 +67,8 @@ func (k Keeper) ModuleBalance(ctx sdk.Context, denom string) sdk.Int {
 // exchange for uTokens. If asset type is invalid or account balance is
 // insufficient, we return an error.
 func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.Coin) error {
-	if !k.IsAcceptedToken(ctx, loan.Denom) {
-		return sdkerrors.Wrap(types.ErrInvalidAsset, loan.String())
+	if err := k.RequireLendEnabled(ctx, loan.Denom); err != nil {
```
Comment (adamewozniak) [inline x/leverage/keeper/keeper.go:70]: I think using the boolean logic is a bit cleaner than having a function return an error if false: 
```suggestion
	if !k.IsLendingEnabled(ctx, loan.Denom) {...
```

File: x/leverage/keeper/keeper.go [hunk 10]
```diff
@@ -220,8 +220,8 @@ func (k Keeper) BorrowAsset(ctx sdk.Context, borrowerAddr sdk.AccAddress, borrow
 		return sdkerrors.Wrap(types.ErrInvalidAsset, borrow.String())
 	}
 
-	if !k.IsAcceptedToken(ctx, borrow.Denom) {
-		return sdkerrors.Wrap(types.ErrInvalidAsset, borrow.String())
+	if err := k.RequireBorrowEnabled(ctx, borrow.Denom); err != nil {
```
Comment (adamewozniak) [inline x/leverage/keeper/keeper.go:223]: Same here:
```suggestion
	if !k.IsBorrowEnabled(ctx, borrow.Denom) {...
```

File: x/leverage/keeper/keeper.go [hunk 11]
```diff
@@ -375,9 +371,6 @@ func (k Keeper) SetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
 
 // GetCollateralSetting checks if a uToken denom is enabled for use as collateral by a single borrower.
 func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddress, denom string) bool {
-	if !k.IsAcceptedUToken(ctx, denom) {
-		return false
-	}
```
Comment (adamewozniak) [inline x/leverage/keeper/keeper.go:380]: Why was this removed?

File: x/leverage/keeper/keeper.go [hunk 12]
```diff
@@ -400,6 +393,9 @@ func (k Keeper) LiquidateBorrow(
 	if !desiredRepayment.IsValid() {
 		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrInvalidAsset, desiredRepayment.String())
 	}
+	if err := k.RequireNotBlacklisted(ctx, desiredRepayment.Denom); err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
```
Comment (adamewozniak) [inline x/leverage/keeper/keeper.go:397]:
```suggestion
	if !k.IsBlacklisted(ctx, desiredRepayment.Denom) {
		return sdk.ZeroInt(), sdk.ZeroInt(), fmt.Errorf...
```

File: x/oracle/keeper/hooks_test.go (hunk 3 above)
Comment (toteki) [reply x/oracle/keeper/hooks_test.go:19]: Ok - in that case, we can remove the TODO for removing oracle accepted assets as a result of leverage module blacklisting.

Would it be ok to make a separate umee repo issue to cover "update of SymbolDenom / Exponent" situation that was mentioned?

**Review by adamewozniak — COMMENTED** [review]
left a few suggestions 🥳

File: x/oracle/keeper/hooks_test.go [hunk 13]
```diff
@@ -9,6 +9,12 @@ func (s *IntegrationTestSuite) TestHooks_AfterTokenRegistered() {
 	h := s.app.OracleKeeper.Hooks()
 	s.Require().Len(s.app.OracleKeeper.AcceptList(s.ctx), 1)
 
+	// TODO: These hooks need to respond to the Token.Blacklist field
+	// (as well as update symbol denom and exponent if they ever change).
+	// Blacklisting in particular should actually eliminate the base
+	// denom from the oracle (and un-blacklisting, for whatever reason,
+	// should do the opposite.)
+
```
Comment (toteki) [inline x/oracle/keeper/hooks_test.go:17]:
```suggestion
```

File: x/leverage/keeper/keeper.go (hunk 9 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go:70]: I started with that function signature actually, but got persuaded by two points:
- I was immediately turning around to make an error "lending not enabled" on seeing the boolean
- It was no longer possible to distinguish "token does not exit" with "lending disabled" on boolean return

File: x/oracle/keeper/hooks_test.go (hunk 3 above)
Comment (adamewozniak) [reply x/oracle/keeper/hooks_test.go]: Imo that's a similar scenario, they could just be updated via param change proposals

File: x/leverage/keeper/keeper.go (hunk 11 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go:380]: Redundant check mostly - a collateral setting would only be true if an asset was acceptable when it was enabled.

In the cases where the asset is no longer accepted (specifically lending disabled or blacklisted now) retuning true if a 0x01 was stored in state doesn't seem harmful.

File: x/leverage/gov_handler_test.go (hunk 7 above)
Comment (toteki) [reply x/leverage/gov_handler_test.go:112]: no tokens were deleted - I'll clarify comment in a different way

File: x/leverage/keeper/keeper.go (hunk 9 above)
Comment (adamewozniak) [reply x/leverage/keeper/keeper.go:70]: Gotcha, makes sense with multiple codepaths - I'd suggest renaming this function to something like `[Check/Validate]LendingEnabled` rather than use the `Required` syntax then 😄

File: x/leverage/gov_handler_test.go (hunk 4 above)
Comment (toteki) [reply x/leverage/gov_handler_test.go:142]: Part of that test exists in types/token_test.go - it ensures that token.Validate fails on that condition.

Also just added a test in this file that ensures the gov proposal fails on invalid token.

File: x/leverage/gov_handler_test.go [hunk 14]
```diff
@@ -30,30 +30,93 @@ func TestUpdateRegistryProposalHandler(t *testing.T) {
...
+					LiquidationThreshold: sdk.MustNewDecFromStr("0.25"),
+					BaseBorrowRate:       sdk.MustNewDecFromStr("0.02"),
+					KinkBorrowRate:       sdk.MustNewDecFromStr("0.22"),
+					MaxBorrowRate:        sdk.MustNewDecFromStr("1.52"),
+					KinkUtilizationRate:  sdk.MustNewDecFromStr("0.8"),
+					LiquidationIncentive: sdk.MustNewDecFromStr("0.1"),
+					EnableLend:           true,
+					EnableBorrow:         true,
+					Blacklist:            false,
+				},
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/gov_handler_test.go]: added to `gov_handler_test.go` - might do a more robust one in `keeper_test.go` next

File: proto/umee/leverage/v1/leverage.proto [hunk 15]
```diff
@@ -117,4 +117,22 @@ message Token {
   // list of allowed tokens.
   string symbol_denom = 10 [(gogoproto.moretags) = "yaml:\"symbol_denom\""];
   uint32 exponent     = 11 [(gogoproto.moretags) = "yaml:\"exponent\""];
+  
+  // Allows lending of this token. Note that withdrawing is always enabled.
+  // Disabling lending would be one step in phasing out an asset type.
+  bool enable_lend = 12;
```
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:123]: See comment on enable_borrow below - if staking derivatives were registered with lending enabled and borrowing disabled, then we would have created a working `meToken`

[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:123]: With staking derivatives, you would stake `sATOM` and lend `sATOM`.... unless we do something more sophisticated.

[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:123]: My impression was that we would not allow meTokens to be borrowed - but they can be used as collateral. That would be one implementation enabled by keeping those booleans separate.

File: CHANGELOG.md [hunk 16]
```diff
@@ -46,6 +46,11 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ## [Unreleased]
 
+### State Machine Breaking
+
+- [913](https://github.com/umee-network/umee/pull/913) Add LendEnabled, BorrowEnabled, and Blacklist to Token struct.
```
[the remark this answers is not in the document]
Comment (toteki) [reply CHANGELOG.md:51]: Done

File: proto/umee/leverage/v1/leverage.proto (hunk 15 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:123]: but here you have `lend` and `borrow`. lend != collateral.
IMO, if an asset is eligible for lending then it's eligible for borrowing.

[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:123]: okay, so mainly a naming issue. Either way, this one needs to allow `MsgLend` and `MsgSetCollateral`, and the other needs to allow `MsgBorrow`

File: x/leverage/keeper/keeper.go (hunk 9 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go:70]: _Note: Resolved the other two conversations to group them here - whatever conclusion we reach here will apply to RequireBorrowEnabled and RequireNotBlacklisted_

File: proto/umee/leverage/v1/leverage.proto (hunk 15 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:123]: So let's rename. This must be clear for users and devs. I see 2 flags:
1. enabled for lending & borrowing
2. enabled as a collateral
`disabled` is implicit = it requires that the 2 flag above are set to false.

File: proto/umee/leverage/v1/leverage.proto [hunk 17]
```diff
@@ -117,4 +117,22 @@ message Token {
...
+  // only, like meTokens.
+  bool enable_borrow = 13;
+
+  // This should only be used to eliminate an asset completely. A blacklisted
+  // asset is treated as though its oracle price is zero, and thus ignored by
+  // calculations such as collateral value and borrow limit. Can still be repaid
+  // or withdrawn, but not liquidated. A blacklisted token must have enable_lend
+  // and enable_borrow set to false. Such tokens can be safely removed from the
+  // by the oracle and price feeder as well.
+  bool blacklist = 14;
```
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:137]: Runs the risk of being interpreted as just a combination of the other two being disabled.

This is more serious in that it also forces the value of the coin to zero in all calculations, which would have all kinds of consequences - we might look for a name that communicates that, but `blacklist` at least shows the severity.

[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:137]: In https://github.com/umee-network/umee/pull/913#discussion_r875948663 I was thinking to remove this field. 

What it would take to unblackist an asset?

[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:137]: Technically, only a governance proposal with the `blacklisted` field set back to false.

In reality, it seems unlikely that an accepted asset could become so nasty we have to blacklist it, but then recover to a point where we need to reverse the decision.

File: proto/umee/leverage/v1/leverage.proto (hunk 15 above)
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:123]: Would changing `(EnableLend, EnableBorrow)` to `(EnableSupply, EnableBorrow)` be better?

One param allows Lend & Collateral actions, and the other allows borrowing.

And separately from this, an asset can also have collateral weight equal to zero (that covers the "borrow and lend, but not collateral" case.)

[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:123]: I don't understand why you want to have `EnableSupply = Lend & Collateral`?
I think it should be:
* EnableCollateral = asset enabled as a collateral for borrowing
* EnableLending = asset enabled for borrowing and lending

[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/leverage.proto:123]: Only `uTokens` can be collateral, and the only way to get a `uToken` is by `MsgLend`

File: x/leverage/keeper/keeper.go (hunk 9 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go]: Renamed all to assert - resolving conversation.

File: proto/umee/leverage/v1/leverage.proto [hunk 18]
```diff
@@ -117,4 +117,22 @@ message Token {
   // list of allowed tokens.
   string symbol_denom = 10 [(gogoproto.moretags) = "yaml:\"symbol_denom\""];
   uint32 exponent     = 11 [(gogoproto.moretags) = "yaml:\"exponent\""];
+  
+  // Allows lending of this token. Note that withdrawing is always enabled.
+  // Disabling lending would be one step in phasing out an asset type.
```
Comment (robert-zaremba) [inline proto/umee/leverage/v1/leverage.proto:122]:
```suggestion
  // Allows lending and setting a collateral using this token. Note that withdrawing 
  // is always enabled. Disabling lending would be one step in phasing out an asset type.
```

File: proto/umee/leverage/v1/leverage.proto [hunk 19]
```diff
@@ -117,4 +117,22 @@ message Token {
...
+  bool enable_lend = 12;
+
+  // Allows borrowing of this token. Note that repaying is always enabled.
+  // Disabling borrowing would be one step in phasing out an asset type, but
+  // could also be used from the start for asset types meant to be collateral
+  // only, like meTokens.
+  bool enable_borrow = 13;
```
Comment (robert-zaremba) [inline proto/umee/leverage/v1/leverage.proto:129]:
```suggestion
  bool enable_msg_borrow = 13;
```

File: proto/umee/leverage/v1/leverage.proto (hunk 15 above)
Comment (robert-zaremba) [inline proto/umee/leverage/v1/leverage.proto:123]:
```suggestion
  // Allows lending of this token. Note that withdrawing is always enabled.
  // Disabling lending would be one step in phasing out an asset type.
  bool enable_msg_lend = 12;
```

**Review by robert-zaremba — APPROVED** [review]
LGTM. Pending proto comments update

File: proto/umee/leverage/v1/leverage.proto [hunk 20]
```diff
@@ -117,4 +117,22 @@ message Token {
   // list of allowed tokens.
   string symbol_denom = 10 [(gogoproto.moretags) = "yaml:\"symbol_denom\""];
   uint32 exponent     = 11 [(gogoproto.moretags) = "yaml:\"exponent\""];
+  
+  // Allows lending and setting a collateral using this token. Note that withdrawing 
+  // is always enabled. Disabling lending would be one step in phasing out an asset type.
+  bool enable_lend = 12;
```
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:123]:
```suggestion
  bool enable_msg_lend = 12;
```
